python-impl/tests.py

```diff
@@ -342,6 +345,74 @@ def test2():
     sk2 = sk
 
 
+def test_threshold_instance(T, N):
+    commitments = []
+    # fragments[i][j] = fragment held by player i,
+    #                   received from player j
+    fragments = [[None] * N for _ in range(N)]
+    secrets = []
+
+    # Step 1 : PrivateKey.new_threshold
+    for player in range(N):
+        secret_key, commi, frags = PrivateKey.new_threshold(T, N)
+        for target, frag in enumerate(frags):
+            fragments[target][player] = frag
+        commitments.append(commi)
+        secrets.append(secret_key)
+
+    # Step 2 : Threshold.verify_secret_fragment
+    for player_source in range(1, N+1):
+        for player_target in range(1, N+1):
+            assert Threshold.verify_secret_fragment(
+                T, fragments[player_target - 1][player_source - 1],
+                player_target, commitments[player_source - 1])
+
+    # Step 3 : master_pubkey = BLS.aggregate_pub_keys(...)
+    #          secret_share = BLS.aggregate_priv_keys(...)
+    master_pubkey = BLS.aggregate_pub_keys(
+           [PublicKey.from_g1(cpoly[0].to_jacobian())
+            for cpoly in commitments],
+           False)
+
+    secret_shares = [BLS.aggregate_priv_keys(map(PrivateKey, row), None, False)
+                     for row in fragments]
+
+    master_privkey = BLS.aggregate_priv_keys(secrets, None, False)
+    msg = 'Test'
+    signature_actual = master_privkey.sign(msg)
+
+    # Step 4 : sig_share = secret_share.sign_threshold(...)
+    # Check every combination of T players
+    for X in combinations(range(1, N+1), T):
+        # X: a list of T indices like [1, 2, 5]
+
+        # Check underlying secret key is correct
+        r = Threshold.interpolate_at_zero(X,
+                [secret_shares[x-1].value for x in X])
+        secret_cand = PrivateKey(r)
+        assert secret_cand == master_privkey
+
+        # Check signatures
+        signature_shares = [secret_shares[x-1].sign_threshold(msg, x, X)
```

Comment:
Weren't we going to do the Lagrange interpolation at aggregation time?

Reply:
Oh I see it's implemented in c++. i think we should have it for python too

Reply:
done